### PR TITLE
pkg/derivation: Use unsafe to cast from []byte to string

### PR DESCRIPTION
**Pull request by adisbladis**
This ensures no copying.

The following test program:
``` go
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"testing"
	"time"

	"github.com/nix-community/go-nix/pkg/derivation"
)

func main() {

	drvPath := "/nix/store/lfb109iifgph2xa94aps13n2w8bpbqsq-texlive-combined-full-2021-final.drv"

	n := 50

	derivationFile, err := os.Open(drvPath)
	if err != nil {
		panic(err)
	}

	drvBytes, err := io.ReadAll(derivationFile)
	if err != nil {
		panic(err)
	}

	{
		start := time.Now()

		for i := 0; i < n; i++ {
			_, err := derivation.ReadDerivation(bytes.NewReader(drvBytes))
			if err != nil {
				panic(err)
			}
		}

		fmt.Println(time.Since(start) / time.Duration(n))
	}

	fmt.Println("Allocs:", int(testing.AllocsPerRun(n, func() {
		_, err := derivation.ReadDerivation(bytes.NewReader(drvBytes))
		if err != nil {
			panic(err)
		}
	})))

}
```

Yields these numbers:
- Before:
```
13.011372ms
Allocs: 27305
```

- After:
```
12.119679ms
Allocs: 19537
```

### CODE REVIEW

**Comment by zimbatm** [comment]
Are you 100% sure the []byte always contains only UTF-8?

**Comment by adisbladis** [comment]
> Are you 100% sure the []byte always contains UTF-8?

That's a good point. I think realistically it always does but I've added a few test cases for other expressions with different encodings.

These actually fail when applied on master but passes here, presumably because on master these are erroneously decoded as utf-8 but here we'll just treat the bytes as utf-8 regardless.

**Comment by flokli** [comment]
This might be git playing tricks - we should probably mark everything inside `test/testdata` as binary, to ensure any (wrong) binary encoding gets preserved.

@adisbladis can you edit `/.gitattributes` and see if this changes things?

**Comment by adisbladis** [comment]
> This might be git playing tricks - we should probably mark everything inside `test/testdata` as binary, to ensure any (wrong) binary encoding gets preserved.
> 
> @adisbladis can you edit `/.gitattributes` and see if this changes things?

Done.

File: .gitattributes [hunk 1]
```diff
@@ -1,3 +1,4 @@
-# Mark json files inside test/testdata as binary,
+# Mark everything inside test/testdata as binary,
 # so git won't replace \n with \r\n on windows.
-test/testdata/*.json binary
+test/testdata/* binary
+test/testdata/*.go text diff
```
Comment (flokli) [inline .gitattributes:4]: This means if we ever embed `build-fixtures.go` in a test fixture itself, it'll look different on windows and linux :thinking: 

Maybe we should use `binary diff` here, to have git keep the file as-is, but still use `diff` to show differences?

Comment (adisbladis) [reply .gitattributes]: Sure thing.

**Pull request merged**